x/compliance/keeper/msg_server_update_compliance_info.go

```diff
@@ -0,0 +1,142 @@
+		return nil, types.NewErrComplianceInfoDoesNotExist(msg.Vid, msg.Pid, msg.SoftwareVersion, msg.CertificationType)
+	}
+
+	if msg.CDVersionNumber != "" {
+		cdVersionNumber, err := strconv.ParseUint(msg.CDVersionNumber, 10, 32)
+
+		if err != nil {
+			return nil, err
+		}
+
+		modelVersion, isFound := k.modelKeeper.GetModelVersion(ctx, msg.Vid, msg.Pid, msg.SoftwareVersion)
+
+		if !isFound {
+			return nil, modeltypes.NewErrModelVersionDoesNotExist(msg.Vid, msg.Pid, msg.SoftwareVersion)
+		}
+
+		if modelVersion.CdVersionNumber != int32(cdVersionNumber) {
+			return nil, types.NewErrModelVersionCDVersionNumberDoesNotMatch(msg.Vid, msg.Pid, msg.SoftwareVersion, msg.CDVersionNumber)
+		}
+
+		complianceInfo.CDVersionNumber = uint32(cdVersionNumber)
+	}
+
+	if msg.CertificationIdOfSoftwareComponent != "" {
+		complianceInfo.CertificationIdOfSoftwareComponent = msg.CertificationIdOfSoftwareComponent
+	}
+
+	if msg.CertificationRoute != "" {
+		complianceInfo.CertificationRoute = msg.CertificationRoute
+	}
+
+	if msg.CompliantPlatformUsed != "" {
+		complianceInfo.CompliantPlatformUsed = msg.CompliantPlatformUsed
+	}
+
+	if msg.CompliantPlatformVersion != "" {
+		complianceInfo.CompliantPlatformVersion = msg.CompliantPlatformVersion
+	}
+
+	if msg.Date != "" {
+		complianceInfo.Date = msg.Date
+	}
+
+	if msg.FamilyId != "" {
+		complianceInfo.FamilyId = msg.FamilyId
+	}
+
+	if msg.OSVersion != "" {
+		complianceInfo.OSVersion = msg.OSVersion
+	}
+
+	if msg.ParentChild != "" {
+		complianceInfo.ParentChild = msg.ParentChild
+	}
+
+	if msg.ProgramType != "" {
+		complianceInfo.ProgramType = msg.ProgramType
+	}
+
+	if msg.ProgramTypeVersion != "" {
+		complianceInfo.ProgramTypeVersion = msg.ProgramTypeVersion
+	}
+
+	if msg.Reason != "" {
+		complianceInfo.Reason = msg.Reason
+	}
+
+	if msg.SupportedClusters != "" {
+		complianceInfo.SupportedClusters = msg.SupportedClusters
+	}
+
+	if msg.Transport != "" {
+		complianceInfo.Transport = msg.Transport
+	}
+
+	// if cdCertificateId is present, update all related indices as well.
+	//nolint:nestif
+	if msg.CDCertificateId != "" {
+		deviceSoftwareCompliance, isFound := k.GetDeviceSoftwareCompliance(ctx, complianceInfo.CDCertificateId)
+
+		index, found := deviceSoftwareCompliance.IsComplianceInfoExist(msg.Vid, msg.Pid, msg.SoftwareVersion)
+		if found {
+			deviceSoftwareCompliance.RemoveComplianceInfo(index)
+		}
+
+		if !isFound {
+			deviceSoftwareCompliance.CDCertificateId = msg.CDCertificateId
+			deviceSoftwareCompliance.ComplianceInfo = append(deviceSoftwareCompliance.ComplianceInfo, &complianceInfo)
+		}
+
+		if len(deviceSoftwareCompliance.ComplianceInfo) == 0 {
+			k.RemoveDeviceSoftwareCompliance(ctx, deviceSoftwareCompliance.CDCertificateId)
+		} else {
+			k.SetDeviceSoftwareCompliance(ctx, deviceSoftwareCompliance)
+		}
+
+		complianceInfo.CDCertificateId = msg.CDCertificateId
+
+		deviceSoftwareCompliance, isFound = k.GetDeviceSoftwareCompliance(ctx, complianceInfo.CDCertificateId)
```

Comment:
Please make sure that we have the following unit test:
1) We have a Compliance Info for Model 1 with CDCertificateId=X1
2) We have a Compliance Info for Model 2 with CDCertificateId=X2
3) We update  Compliance Info for Model 2 to be CDCertificateId=X1
4) Check that query by CDCertificateId=X2 returns nothing (not found)
5) Check that query by CDCertificateId=X1 returns a list of two values (for model 1 and for model 2)